Add chorus, reverb dry and master gain for sonivoxrender

rsp4jack opened the pull request:
reverb wet is changed to 0..32767 from 0..32765

rsp4jack commented:
I have updated `sonivoxrender.1.md`, but pandoc does not work for me, then I left `sonivoxrender.1` unchanged.

pedrolcl commented:
> I have updated `sonivoxrender.1.md`, but pandoc does not work for me, then I left `sonivoxrender.1` unchanged.

To build the man page there is an additional cmake option that you need to set: `BUILD_MANPAGE=Yes`. 

The build system does not overwrite an existing man page that has a newer timestamp than its source (like any other build artifact) so it is better to start with an empty build directory or delete the old manpage in the build directory.

rsp4jack commented:
Maybe man pages should be built in CI?

pedrolcl commented:
ea38ba3 is wrong. Please revert.

The **source tree** should always contain the latest released or updated man page. You may delete the man page in the **build tree**, though.

> Maybe man pages should be built in CI?

That would require to install `pandoc` as a dependency on the CI host. This is a **huge** dependency, that may  [be unavailable](https://github.com/msys2/MINGW-packages/issues/12485) for all configurations, and I prefer to avoid it.

rsp4jack commented:
Sorry for misunderstood, and I just updated the man pages.

pedrolcl commented:
What about the other chorus parameters?

https://github.com/pedrolcl/sonivox/blob/290e469a8fd4871413e4c544f7820650c36bc056/arm-wt-22k/host_src/eas_chorus.h#L34-L42

IMHO, at leas, the level parameter should be included.

pedrolcl commented on example/sonivoxrender.1.md:
```diff
@@ -30,7 +30,19 @@ It reads .MID (Standard MIDI Files) file format, and writes an audio stream to t
+
+-c  _chorus_preset_
+
+:   Chorus preset between 0 and 4: 0=no, 1..4=presents.
```
Typo:

```suggestion
:   Chorus preset between 0 and 4: 0=no, 1..4=presets.
```

rsp4jack commented:
The chorus presets override the other parameters (rate, depth and level), as they are presets, not chorus types. Have the preset and other parameters both configurable may introduce unwanted complexity. 

https://github.com/pedrolcl/sonivox/blob/290e469a8fd4871413e4c544f7820650c36bc056/arm-wt-22k/lib_src/eas_chorus.c#L585-L604

pedrolcl commented:
> The chorus presets override the other parameters (rate, depth and level), as they are presets, not chorus types. Have the preset and other parameters both configurable may introduce unwanted complexity.

This is a similar case to overriding the reverb parameters:

https://github.com/pedrolcl/sonivox/blob/290e469a8fd4871413e4c544f7820650c36bc056/arm-wt-22k/lib_src/eas_reverb.c#L914-L954

You may try how the interaction between presets and individual parameters works in [/pedrolcl/drumstick](https://github.com/pedrolcl/drumstick)

![image](https://github.com/user-attachments/assets/44c0c9cf-b5f0-403d-90af-4524094840e4)

rsp4jack commented:
I just came up an idea, to add a pseudo "custom preset", to enable users to edit other parameters.
This will work fine in chorus, but reverb have tons of parameters which is private.

rsp4jack commented:
BTW currently sonivoxrender's reverb wet setting works only for preset 1 (begin from 0), as setting it to other presets will override the wet setting.

pedrolcl commented:
> I just came up an idea, to add a pseudo "custom preset", to enable users to edit other parameters. This will work fine in chorus, but reverb have tons of parameters which are private.

You don't need that if you simply apply the preset setting **before** setting the the overrides.

pedrolcl commented:
> BTW currently sonivoxrender's reverb wet setting works only for preset 2 (begin from 0), as setting it to other presets will override the wet setting.

Current 'devel' branch fixes this isue

rsp4jack commented:
> You don't need that if you simply apply the preset setting **before** setting the overrides.

The preset setting only applies when `ChorusUpdate` is called, which is called only when mixing audio. https://github.com/pedrolcl/sonivox/blob/290e469a8fd4871413e4c544f7820650c36bc056/arm-wt-22k/lib_src/eas_chorus.c#L282

rsp4jack commented:
I rebased to the current `devel` branch, and refactored `Chorus/ReverbSetParam` to apply presets immediately. Now it should work, just like you said

> simply apply the preset setting before setting the overrides.

And added chorus level CLI option for sonivoxrender.

pedrolcl commented on arm-wt-22k/lib_src/eas_chorus.c:
```diff
@@ -590,9 +557,19 @@ static EAS_RESULT ChorusUpdate(S_CHORUS_OBJECT *pChorusData)
     pChorusData->m_nRate =  pPreset->m_nRate;
     pChorusData->m_nDepth = pPreset->m_nDepth;
 
+    //chorus rate and depth need some massaging from preset value (which is sample rate independent)
+
+    //convert rate from steps of .05 Hz to value which can be used as phase increment,
+    //with current CHORUS_SHAPE_SIZE and rate limits, this fits into 16 bits
+    //want to compute ((shapeSize * 65536) * (storedRate/20))/sampleRate;
+    //computing it as below allows rate steps to be evenly spaced
+    //uses 32 bit divide, but only once when new value is selected
     pChorusData->m_nRate = (EAS_I16)
         ((((EAS_I32)CHORUS_SHAPE_SIZE<<16)/(20*(EAS_I32)_OUTPUT_SAMPLE_RATE)) * pChorusData->m_nRate);
 
+    //convert depth from steps of .05 ms, to samples, with 16 bit whole part, discard fraction
+    //want to compute ((depth * sampleRate)/20000)
+    //use the following approximation since 105/32 is roughly 65536/20000
```
I'm glad that you preserved the comments about the hacks in these calculations! Nice!

pedrolcl commented on arm-wt-22k/lib_src/eas_reverbdata.h:
Here we can also do some more cleanup, for instance remove L360:

https://github.com/rsp4jack/sonivox/blob/1472c2b7ab88ff6a2d468bbe56ddc4092cbd2806/arm-wt-22k/lib_src/eas_reverbdata.h#L360

pedrolcl commented on arm-wt-22k/lib_src/eas_reverb.c:
More cleanup is possible. Remove L79 and L82:

https://github.com/rsp4jack/sonivox/blob/1472c2b7ab88ff6a2d468bbe56ddc4092cbd2806/arm-wt-22k/lib_src/eas_reverb.c#L79-L82

pedrolcl commented on example/sonivoxrender.c:
```diff
@@ -101,11 +106,18 @@ int initializeLibrary(void)
-    if ( preset >= EAS_PARAM_REVERB_LARGE_HALL && preset <= EAS_PARAM_REVERB_ROOM ) {
-        sw = EAS_FALSE;
-        result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_REVERB, EAS_PARAM_REVERB_PRESET, preset);
+    result = EAS_SetVolume(mEASDataHandle, NULL, playback_gain);
+    if (result != EAS_SUCCESS) {
+        fprintf(stderr, "Failed to set volume\n");
+        ok = EXIT_FAILURE;
+        goto cleanup;
+    }
```
Although the function is called EAS_SetVolume() and the value unit is is dB, it should be called gain (as your local variable suggest) in the context of this program (and probably on any other as well) .

pedrolcl commented on example/sonivoxrender.c:
```diff
@@ -283,8 +333,36 @@ int main (int argc, char **argv)
+                return EXIT_FAILURE;
+            }
+            break;
+        case 'v':
```
The option 'v' should be reserved for the program version, like in most other Unix command line programs: https://tldp.org/LDP/abs/html/standard-options.html
The gain option could be called 'g' instead.

pedrolcl commented on example/sonivoxrender.c:
```diff
@@ -259,17 +304,22 @@ int main (int argc, char **argv)
     opterr = 0;
 
-    while ((c = getopt (argc, argv, "hd:r:w:")) != -1) {
+    while ((c = getopt (argc, argv, "hd:r:w:n:c:l:v:")) != -1) {
```
Maybe now it would be a good opportunity to provide long option names:
https://www.gnu.org/software/libc/manual/html_node/Getopt-Long-Options.html

rsp4jack replied:
Nice, but I think that would be too much for this PR. Maybe later.

pedrolcl replied:
OK, no problem.

pedrolcl reviewed (commented):
Good job with your refactoring! I've added some additional suggestions.
